polynote-frontend/polynote/cell.js

```diff
@@ -378,49 +385,68 @@ export class CodeCell extends Cell {
 
         const rel = args.rel || 'none';
         const lang = args.lang || null;
-        const self = this;
         return CodeCell.parseContent(content, mimeType, lang).then(function(result) {
-            const el = div(['output'], result).attr('rel', rel).attr('mime-type', mimeType);
-            self.cellOutputDisplay.appendChild(el);
-            return el;
+            return div(['output'], result).attr('rel', rel).attr('mime-type', mimeType);
         }).catch(function(err) {
-            self.cellOutputDisplay.appendChild(
-                div(['output'], err)
-            );
+            return div(['output'], err);
         });
     }
 
+    addOutput(contentType, content) {
+        this.cellOutputDisplay.classList.add('output');
+        if (!this.container.classList.contains('error')) {
+            this.container.classList.add('success');
+        }
+        const self = this;
+        this.buildOutput(contentType, content).then(function(el) {
+            self.cellOutputDisplay.appendChild(el);
+        })
+    }
+
     addResult(result) {
         if (result instanceof ResultValue) {
+            // clear results
+            this.resultTabs.innerHTML = '';
+
             // TODO: keep "result" and "output" separate for UI... have a way to show declarations, results, outputs, etc. separately
             if (result.name !== "Out") {
                 // don't display this; it's a named declaration
                 // TODO: have a way to display these if desired
-            } else if (!this.resultEl || !this.resultEl.parentNode || !this.resultEl.reprs || !this.resultEl.reprs.length || this.resultEl.reprs.length < result.reprs.length) {
-                if (this.resultEl && this.resultEl.parentNode) {
-                    this.resultEl.parentNode.removeChild(this.resultEl);
-                }
-                this.resultEl = null;
-                const [mime, content] = result.displayRepr;
-                const outLabel = result.reprs.length <= 1
-                    ? div(['out-ident'], `Out:`)
-                    : div(['out-ident', 'with-reprs'], `Out:`).click(evt => {
-                        const reprUi = new ReprUI(`Cell${this.id}`, this.path, result.reprs, this.resultEl);
-                        reprUi.setEventParent(this);
-                        reprUi.show();
-                    });
+            } else if (result.reprs.length > 1) {
+                const outLabel = div(['out-ident', 'with-reprs'], `Out:`);
+                this.resultTabs.appendChild(outLabel);
 
-                this.addOutput(mime, content).then(el => {
-                    el.reprs = result.reprs;
-                    this.resultEl = el;
-                    el.insertBefore(outLabel, el.childNodes[0])
+                const [mime, content] = result.displayRepr;
+                const self = this;
+                this.buildOutput(mime, content).then(function(el) {
```

Comment:
Note: if you use the arrow lambda syntax `el => {...}` then you get to keep `this` and you don't have to rebind it to `self`.

Reply:
Good to know - thanks!